rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,365 @@
+  ...shorthands.borderRadius(0),
+  opacity: 1,
+  transform: 'translate3D(0, 0, 0) scale(1)',
+};
+
+const hiddenKeyframe = {
+  ...shorthands.borderRadius('36px'),
+  opacity: 0,
+  transform: 'translate3D(-100%, 0, 0) scale(0.9)',
+};
+
+const useStyles = makeStyles({
+  root: {
+    willChange: 'opacity, transform, border-radius',
+  },
+
+  entering: {
+    animationDuration: '500ms',
+    animationTimingFunction: tokens.curveDecelerateMid,
+    animationName: {
+      '0%': hiddenKeyframe,
+      '100%': visibleKeyframe,
+    },
+  },
+
+  exiting: {
+    animationDuration: '750ms',
+    animationTimingFunction: tokens.curveAccelerateMin,
+    animationName: {
+      '0%': visibleKeyframe,
+      '100%': hiddenKeyframe,
+    },
+  },
+});
+
+export const useSampleStyles_unstable = (state: SampleState): SampleState => {
+  const styles = useStyles();
+
+  state.root.className = mergeClasses(
+    SampleClassNames.root,
+    state.motion.type === 'entering' && styles.entering,
+    state.motion.type === 'exiting' && styles.exiting,
+    styles.root,
+  );
+
+  return state;
+};
+```
+
+#### Overriding the transition/animation
+
+To override motion on another component, only a CSS change is needed:
+
+```tsx
+import * as React from 'react';
+import { Drawer } from '@fluentui/react-drawer';
+import { makeStyles } from '@fluentui/react-components';
+
+const useStyles = makeStyles({
+  customDuration: {
+    transitionDuration: '500ms',
+  },
+});
+
+export const CustomDuration = () => {
+  const styles = useStyles();
+
+  return <Drawer className={styles.customDuration} />;
+};
+```
+
+In order to enable overrides, Fluent components can also accept a prop that can be used to receive `useMotion` values coming from another component. This is extremely useful for cases when a completely custom animation/transition is needed. In this case, we enable full control to override animations/transitions while improving performance by not computing the values twice. In the following example, the `open` prop for the Drawer can receive either a `boolean` or a `MotionState`:
+
+##### Application side:
+
+```tsx
+import * as React from 'react';
+import { makeStyles, mergeClasses, Drawer, Button } from '@fluentui/react-components';
+import { useMotion } from '@fluentui/react-motion-preview';
+
+const useStyles = makeStyles({
+  drawer: {
+    opacity: 0,
+    transitionDuration: '3s',
+    transitionProperty: 'opacity',
+  },
+
+  drawerVisible: {
+    opacity: 1,
+  },
+});
+
+export const CustomAnimation = () => {
+  const styles = useStyles();
+
+  const [open, setOpen] = React.useState(false);
+  const motion = useMotion<HTMLDivElement>(open);
+
+  const onClick = () => setOpen(!open);
+
+  return (
+    <div className={styles.root}>
+      <Button appearance="primary" onClick={onClick}>
+        Toggle
+      </Button>
+      <Drawer open={motion} className={mergeClasses(styles.drawer, motion.isActive() && styles.drawerVisible)} />;
```

Comment:
is it intentional that the motion override uses the `open` prop?

Reply:
This is an example on a possible approach for it, as useMotion can receive both a boolean or a MotionState.